lib/puppet/module/plan.rb

```diff
@@ -0,0 +1,161 @@
+      def initialize(message, kind, details = nil)
+        super(message)
+        @details = details || {}
+        @kind = kind
+      end
+
+      def to_h
+        {
+          msg: message,
+          kind: kind,
+          details: details
+        }
+      end
+    end
+
+    class InvalidName < Error
+      def initialize(name, msg)
+        super(msg, 'puppet.plans/invalid-name')
+      end
+    end
+
+    class InvalidFile < Error
+      def initialize(msg)
+        super(msg, 'puppet.plans/invalid-file')
+      end
+    end
+
+    class InvalidPlan < Error
+    end
+    class InvalidMetadata < Error
+    end
+    class PlanNotFound < Error
+      def initialize(plan_name, module_name)
+        msg = _("Plan %{plan_name} not found in module %{module_name}.") %
+          {plan_name: plan_name, module_name: module_name}
+        super(msg, 'puppet.plans/plan-not-found', { 'name' => plan_name })
+      end
+    end
+
+    ALLOWED_EXTENSIONS = %w{.pp .yaml}
+    RESERVED_WORDS = %w{and application attr case class consumes default else
+        elsif environment false function if import in inherits node or private
+        produces site true type undef unless}
+    RESERVED_DATA_TYPES = %w{any array boolean catalogentry class collection
+        callable data default enum float hash integer numeric optional pattern
+        resource runtime scalar string struct tuple type undef variant}
+    MOUNTS = %w[lib files plans]
+
+    def self.is_plan_name?(name)
+      return true if name =~ /^[a-z][a-z0-9_]*$/
+      return false
+    end
+
+    # Determine whether a plan file has a legal name and extension
+    def self.is_plans_filename?(path)
+      name = File.basename(path, '.*')
+      ext = File.extname(path)
+      return [false, _("Plan names must start with a lowercase letter and be composed of only lowercase letters, numbers, and underscores")] unless is_plan_name?(name)
+      unless ALLOWED_EXTENSIONS.include? ext
+        return [false, _("Plan name cannot have extension %{ext}, must be .pp or .yaml") % { ext: ext }]
+      end
+      if RESERVED_WORDS.include?(name)
+        return [false, _("Plan name cannot be a reserved word, but was '%{name}'") % { name: name }]
+      end
+      if RESERVED_DATA_TYPES.include?(name)
+        return [false, _("Plan name cannot be a Puppet data type, but was '%{name}'") % { name: name }]
+      end
+      return [true]
+    end
+
+    # Executables list should contain the full path of all possible implementation files
+    def self.find_implementations(name, plan_files)
+      basename = name.split('::')[1] || 'init'
+
+      # If implementations isn't defined, then we use executables matching the
+      # plan name, and only one may exist.
+      implementations = plan_files.select { |impl| File.basename(impl, '.*') == basename }
+
+      # TODO Flagged for review.
+      # Select .pp before .yaml, since .pp comes before .yaml alphabetically.
+      chosen = implementations.sort.first
```

Comment:
Is this a good way to select if both the `.pp` and `.yaml` plan files exist?

Reply:
@puppetlabs/skeletor any insights here? Thanks!

Reply:
Assuming pp and yaml will be the only valid plan filename extensions for the forseeable future, I'd say yeah